pkg/forwarder/forwarder_test.go

```diff
@@ -6,278 +6,153 @@
-	defer config.Datadog.Set("proxy", nil)
-
-	forwarder := NewDefaultForwarder(map[string][]string{targetURL: []string{firstKey}})
-	// delay next retry cycle so we can inspect retry queue
-	flushInterval = 5 * time.Hour
-	forwarder.NumberOfWorkers = 1
 	forwarder.Start()
-	defer forwarder.Stop()
+	err = forwarder.sendHTTPTransactions(tr)
+	assert.Nil(t, err)
+	forwarder.Stop()
+}
 
-	payload := []byte("SubmitSeries payload")
-	payloads := []*[]byte{}
-	payloads = append(payloads, &payload)
-	assert.Nil(t, forwarder.SubmitSeries(payloads, map[string]string{"Content-Type": "application/unit-test"}))
-	// wait for the queries to complete before changing expected value
-	<-wait
-	require.Len(t, forwarder.retryQueue, 0)
+func TestRequeueTransaction(t *testing.T) {
+	forwarder := NewDefaultForwarder(nil)
+	tr := NewHTTPTransaction()
+	assert.Len(t, forwarder.retryQueue, 0)
+	forwarder.requeueTransaction(tr)
+	assert.Len(t, forwarder.retryQueue, 1)
 }
 
-func TestSubmitWithProxyAndPassword(t *testing.T) {
-	targetURL := "http://test:2121"
-	userInfo := "testuser:password123456"
-	expectedAuth := fmt.Sprintf("Basic %s", base64.StdEncoding.EncodeToString([]byte(userInfo)))
-	firstKey := "api_key1"
-	wait := make(chan bool)
-
-	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		defer func() { wait <- true }()
-		assert.Equal(t, "POST", r.Method)
-		assert.Equal(t, targetURL, fmt.Sprintf("%s://%s", r.URL.Scheme, r.URL.Host))
-		assert.Equal(t, firstKey, r.Header.Get(apiHTTPHeaderKey))
-		assert.Equal(t, expectedAuth, r.Header.Get("Proxy-Authorization"))
-
-		w.WriteHeader(http.StatusOK)
-	}))
-	defer ts.Close()
-	config.Datadog.Set("proxy", map[string]interface{}{"http": fmt.Sprintf("http://%s@%s", userInfo, ts.URL[7:])})
-	defer config.Datadog.Set("proxy", nil)
-
-	forwarder := NewDefaultForwarder(map[string][]string{targetURL: []string{firstKey}})
-	// delay next retry cycle so we can inspect retry queue
-	flushInterval = 5 * time.Hour
-	forwarder.NumberOfWorkers = 1
-	forwarder.Start()
-	defer forwarder.Stop()
+func TestRetryTransactions(t *testing.T) {
+	forwarder := NewDefaultForwarder(nil)
+	forwarder.init()
+	forwarder.retryQueueLimit = 1
 
-	payload := []byte("SubmitSeries payload")
-	payloads := []*[]byte{}
-	payloads = append(payloads, &payload)
-	assert.Nil(t, forwarder.SubmitSeries(payloads, map[string]string{"Content-Type": "application/unit-test"}))
-	// wait for the queries to complete before changing expected value
-	<-wait
-	require.Len(t, forwarder.retryQueue, 0)
+	t1 := NewHTTPTransaction()
+	t1.nextFlush = time.Now().Add(-1 * time.Hour)
+	t2 := NewHTTPTransaction()
+	t2.nextFlush = time.Now().Add(1 * time.Hour)
+	forwarder.requeueTransaction(t2)
+	forwarder.requeueTransaction(t2) // this second one should be dropped
+	forwarder.requeueTransaction(t1) // the queue should be sorted
+	forwarder.retryTransactions(time.Now())
+	assert.Len(t, forwarder.retryQueue, 1)
```

Comment:
we forgot asserting `droppedTransaction == 1`

Reply:
Problem is `droppedTransaction` is local to the function and only used there

Reply:
Fixed, accessing expvar directly